crates/pallet-grandpa-finality-verifier/src/lib.rs

```diff
@@ -137,50 +144,59 @@ pub mod pallet {
         object: &[u8],
     ) -> Result<(C::Hash, C::BlockNumber), DispatchError> {
         let block = C::decode_block::<T>(object)?;
-        let justification = block
-            .justifications
-            .ok_or(Error::<T>::MissingJustification)?
-            .into_justification(GRANDPA_ENGINE_ID)
-            .ok_or(Error::<T>::MissingJustification)?;
-        let justification = C::decode_grandpa_justifications::<T>(justification.as_slice())?;
-        let best_finalized = {
-            let data =
-                BestFinalized::<T>::get(chain_id).ok_or(Error::<T>::FinalizedHeaderNotFound)?;
-            C::decode_header::<T>(data.as_slice())
-        }?;
-
-        // ensure block is always increasing
         let (number, hash) = (block.block.header.number(), block.block.header.hash());
-        let next_block_number = best_finalized
-            .number()
-            .checked_add(&One::one())
-            .ok_or(sp_runtime::ArithmeticError::Overflow)?;
-        ensure!(next_block_number == *number, Error::<T>::InvalidHeader);
-
-        // fetch current authority set
-        let authority_set = <CurrentAuthoritySet<T>>::get(chain_id);
-        let voter_set =
-            VoterSet::new(authority_set.authorities).ok_or(Error::<T>::InvalidAuthoritySet)?;
-        let set_id = authority_set.set_id;
-
-        // verify justification
-        verify_justification::<C::Header>((hash, *number), set_id, &voter_set, &justification)
-            .map_err(|e| {
-                log::error!(
-                    target: "runtime::grandpa-finality-verifier",
-                    "Received invalid justification for {:?}: {:?}",
-                    hash,
-                    e,
-                );
-                Error::<T>::InvalidJustification
-            })?;
-
-        // Update any next authority set if any
-        let next_header = &block.block.header;
-        try_enact_authority_change::<T, C>(chain_id, next_header, set_id)?;
+
+        match BestFinalized::<T>::get(chain_id) {
+            None => {
+                // this is the first block
+                // import as is
+                BestFinalized::<T>::insert(chain_id, block.block.header.encode());
+                return Ok((hash, *number));
+            }
+            Some(data) => {
+                let best_finalized = C::decode_header::<T>(data.as_slice())?;
+                // ensure block is always increasing
+                let next_block_number = best_finalized
+                    .number()
+                    .checked_add(&One::one())
+                    .ok_or(sp_runtime::ArithmeticError::Overflow)?;
+                ensure!(next_block_number == *number, Error::<T>::InvalidHeader);
+            }
+        };
+
+        // check if we should validate or just import
+        if (*block.block.header.number()).into() > ValidationCheckpoint::<T>::get(chain_id) {
```

Comment:
```suggestion
        if number.into() > ValidationCheckpoint::<T>::get(chain_id) {
```

Also shouldn't it start validating with the block at the checkpoint too (`>=`)?

I think it is related to `None => {}` branch above, but I don't fully understand how.

Reply:
> Also shouldn't it start validating with the block at the checkpoint too (>=)?

No. the header that contains the Authority set change will be last block signed by previous authority set. So we can't validate that last block with new authority set

Reply:
Interesting, can't we still verify its body though since we should already have the header from initialization?

Reply:
> https://github.com/subspace/subspace/pull/380#discussion_r854467662

Related to this. 
Ofcourse we can force the check where `block_number == 0 ` as well. This just makes it easy to sync from any number instead of 0

Reply:
I'm not following. What do you mean by `block_number == 0`?

I think I get something. We have `BestFinalized` as an option. But can't we make it non-option and store the header from initialization in there from the very beginning?

Reply:
Well, best finalization cannot be filled from init since that header belongs to a descendent.

If the checkpoint is 100th block. But while importing the blocks, we start from 0 for example. 
Then our check to ensure block number is always increasing breaks.
Hence, we set the first header to the feed as best finalized.

Ofcourse there is a case where first header could be above the Checkpoint. In that case, yes we wont do the validation but we also wont update any authority set that header has. So next blocks will always go through the validation logic.
I think I will update to ensure first block to the feed is always <= checkpoint. That should solve issue where first block is after checkpoint.

> I'm not following. What do you mean by block_number == 0?

We can always ensure that first block is genesis by checking block number of the target to 0.